include/NeoFOAM/core/registerClass.hpp

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: MIT
+// SPDX-FileCopyrightText: 2024 NeoFOAM authors
+
+// additional information see:
+// https://stackoverflow.com/questions/52354538/derived-class-discovery-at-compile-time
+// https://stackoverflow.com/questions/10332725/how-to-automatically-register-a-class-on-creation
+
+#pragma once
+
+#include <functional>
+#include <iostream>
+
+namespace NeoFOAM
+{
+
+// Type trait to check if a type is a std::function
+template<typename T>
+struct is_std_function : std::false_type
+{
+};
+
+template<typename Ret, typename... Args>
+struct is_std_function<std::function<Ret(Args...)>> : std::true_type
+{
+};
+
+// Concept that uses the type trait
+template<typename T>
+concept StdFunction = is_std_function<T>::value;
+
+/**
+ * @brief Template struct for managing class registration.
+ *
+ * This struct provides a mechanism for registering classes with a given base class and a create
+ * function. It maintains a map of class names to create functions, allowing for dynamic class
+ * instantiation.
+ *
+ * @tparam baseClass The base class type.
+ * @tparam CreateFunction The type of the create function.
+ */
+template<typename baseClass, StdFunction CreateFunction>
+class RegisterClassManager // needs to be a class otherwise breathe will not document it
+{
+public:
+
+    /**
+     * @brief Registers a class with the given name and create function.
+     *
+     * @param name The name of the class.
+     * @param createFunc The create function for the class.
+     * @return true if the class was successfully registered, false if the class name already exists
+     * in the map.
+     * @throws std::runtime_error if the class name already exists in the map.
+     */
+    static bool registerClass(const std::string name, CreateFunction createFunc)
+    {
+        auto result = classMap.insert({name, createFunc});
+        if (!result.second)
+        {
+            throw std::runtime_error("Insertion failed: Key already exists.");
+        }
+        return result.second;
+    }
+
+    /**
+     * @brief Gets the number of registered classes.
+     *
+     * @return The number of registered classes.
+     */
+    static size_t size() { return classMap.size(); }
```

Comment:
I guess the function can be marked const.
```suggestion
    static size_t size() const { return classMap.size(); }
```